pubsub-mapped-api/src/main/java/com/google/pubsub/clients/producer/KafkaProducer.java

```diff
@@ -0,0 +1,424 @@
+              .build())
+
+          .build();
+    } catch (IOException e) {
+      throw new KafkaException("Failed to construct kafka producer", e);
+    }
+    return pub;
+  }
+
+  /**
+   * Sends the given record.
+   */
+  public Future<RecordMetadata> send(ProducerRecord<K, V> originalRecord) {
+    return send(originalRecord, null);
+  }
+
+  //TODO: there is an onSendError() in the interceptors, while there are no errors.
+
+  /**
+   * Sends the given record and invokes the specified callback.
+   * The given record must have the same topic as the producer.
+   */
+  public Future<RecordMetadata> send(ProducerRecord<K, V> originalRecord, Callback callback) {
+    if (closed.get()) {
+      throw new IllegalStateException("Cannot send after the producer is closed.");
+    }
+
+    ProducerRecord<K, V> record =
+        this.interceptors == null ? originalRecord : this.interceptors.onSend(originalRecord);
+
+    if (record == null) {
+      throw new NullPointerException();
+    }
+
+    DateTime dateTime = new DateTime();
+
+    byte[] valueBytes = null;
+    if (record.value() != null) {
+      try {
+        valueBytes = valueSerializer.serialize(record.topic(), record.value());
+      } catch (ClassCastException e) {
+        throw new SerializationException("Can't convert value of class " +
+            record.value().getClass().getName() + " to class " +
+            producerConfig.getKafkaConfigs().getClass(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG).getName() +
+            " specified in value.serializer");
+      }
+    }
+
+    if (valueBytes == null || valueBytes.length == 0) {
+      throw new NullPointerException("Value cannot be null or an empty string.");
+    }
+
+    byte[] keyBytes = null;
+    if (record.key() != null) {
+      try {
+        keyBytes = keySerializer.serialize(record.topic(), record.key());
+      } catch (ClassCastException cce) {
+        throw new SerializationException("Can't convert key of class " +
+            record.key().getClass().getName() + " to class " +
+            producerConfig.getKafkaConfigs().getClass(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG).getName() +
+            " specified in key.serializer");
+      }
+    }
+
+    PubsubMessage msg = createMessage(record, dateTime.toString(), keyBytes, valueBytes);
+
+    ApiFuture<String> messageIdFuture = publishers.computeIfAbsent(
+            record.topic(), topic -> createPublisher(topic)).publish(msg);
+
+    final Callback cb = callback;
+    final String topic = record.topic();
+    final int keySize = keyBytes == null ? 0 : keyBytes.length, valueSize = valueBytes.length;
+
+    final SettableFuture<RecordMetadata> future = SettableFuture.create();
+
+    if (callback != null) {
+      if (isAcks) {
+        ApiFutures.addCallback(messageIdFuture, new ApiFutureCallback<String>() {
+          private RecordMetadata recordMetadata =
+              getRecordMetadata(topic, dateTime.getMillis(), keySize, valueSize);
```

Comment:
Shouldn't the offset passed into here match the timestamp that was set for the "timestamp" attribute of the message? Probably worth checking in a unit test.

Reply:
Isn't it set by the server side? please elaborate.

Reply:
Actually, I was mistaken. You do use the same dateTime object for the recordMetadata and the timestamp attribute. However, I'd recommend passing the dateTime attribute as something other than a string to createMessage. I think using the milliseconds for it makes the most sense. And in fact, let's rename "timestamp" to "offset." The fact that we are using the timestamp is not necessarily the interesting part. It is that we are providing an offset that is consistent across the publisher and subscriber.

Reply:
Done.